### PR TITLE
[MVPN-239] OS X mysterium client upgrade

### PR DESCRIPTION
**Pull request by donce**

### CODE REVIEW

File: src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js [hunk 1]
```diff
@@ -55,25 +64,58 @@ class LaunchDaemonProcess implements Process {
...
-    await this._monitoring.waitForStatusDownWithTimeout()
 
-    logger.info('Restarting: process is down, starting it up')
-    await this.start()
-    logger.info('Restarting: waiting for process to be up')
-    await this._monitoring.waitForStatusUpWithTimeout()
+    const unsubscribe = this._monitoring.onStatusDown(() => { this.start() })
+    try {
+      logger.info('Upgrading: waiting for upgraded client')
+      await this._waitForUpgrade(15000)
```
[the remark this answers is not in the document]
Comment (mipo47) [reply src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js]: error thrown by _waitForUpgrade is ignored

File: src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js [hunk 2]
```diff
@@ -55,25 +64,58 @@ class LaunchDaemonProcess implements Process {
...
+    setTimeout(() => {
+      timeFinished = true
+    }, timeout)
 
-    logger.info('Restarting: process is up')
+    while (!timeFinished) {
+      if (await this._isUpgradeFinished()) {
+        return
+      }
+      const delay = 200
```
Comment (mipo47) [inline src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js]: Put const out of cycle or don't use it at all

File: src/libraries/mysterium-client/monitoring/monitoring.js [hunk 3]
```diff
@@ -57,65 +58,65 @@ class Monitoring {
...
    */
-  onNewStatusUp (callback: EmptyCallback): void {
-    this._upSubscriber.subscribe(callback)
+  onNewStatusUp (callback: EmptyCallback): Unsubscriber {
+    return this._upSubscriber.subscribe(callback)
   }
 
   waitForNewStatusUpWithTimeout (): Promise<void> {
-    return onFirstEventOrTimeout(this.onNewStatusUp.bind(this), MYSTERIUM_CLIENT_WAITING_THRESHOLD)
+    return onFirstEventOrTimeout((cb) => { this.onNewStatusUp(cb) }, MYSTERIUM_CLIENT_WAITING_THRESHOLD)
```
Comment (mipo47) [inline src/libraries/mysterium-client/monitoring/monitoring.js]: mind-blowing :)

**Comment by donce** [comment]
@mipo47 fixed

File: src/libraries/mysterium-client/monitoring/monitoring.js (hunk 3 above)
Comment (donce) [reply src/libraries/mysterium-client/monitoring/monitoring.js]: Yeah, a bit complex. Any suggestions?

File: src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (donce) [reply src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js]: It is handled at a higher level - if upgrade fails, which is part of start - that means start failed as well.

File: src/libraries/mysterium-client/monitoring/monitoring.js (hunk 3 above)
Comment (shroomist) [reply src/libraries/mysterium-client/monitoring/monitoring.js]: still couldn't get my head around this.
where does `cb` come from?

File: src/libraries/mysterium-client/monitoring/monitoring.js [hunk 4]
```diff
@@ -57,65 +58,65 @@ class Monitoring {
...
   /**
    * Triggers once service is down. Does not trigger instantly if it is already down.
    */
-  onNewStatusDown (callback: EmptyCallback): void {
-    this._downSubscriber.subscribe(callback)
+  onNewStatusDown (callback: EmptyCallback): Unsubscriber {
+    return this._downSubscriber.subscribe(callback)
   }
 
   waitForNewStatusDownWithTimeout (): Promise<void> {
```
Comment (mipo47) [inline src/libraries/mysterium-client/monitoring/monitoring.js]: unused

File: src/libraries/mysterium-client/monitoring/monitoring.js [hunk 5]
```diff
@@ -57,65 +58,65 @@ class Monitoring {
...
-  onStatusDown (callback: EmptyCallback): void {
-    this.onNewStatusDown(callback)
+  onStatusDown (callback: EmptyCallback): Unsubscriber {
     if (this._lastStatus === false) {
       callback()
     }
+    return this.onNewStatusDown(callback)
   }
 
   waitForStatusDownWithTimeout (): Promise<void> {
```
Comment (mipo47) [inline src/libraries/mysterium-client/monitoring/monitoring.js]: no tests for public method

File: src/libraries/mysterium-client/monitoring/monitoring.js [hunk 6]
```diff
@@ -57,65 +58,65 @@ class Monitoring {
...
-  onStatusUp (callback: EmptyCallback): void {
-    this.onNewStatusUp(callback)
+  onStatusUp (callback: EmptyCallback): Unsubscriber {
     if (this._lastStatus) {
       callback()
     }
+    return this.onNewStatusUp(callback)
   }
 
   waitForStatusUpWithTimeout (): Promise<void> {
```
Comment (mipo47) [inline src/libraries/mysterium-client/monitoring/monitoring.js]: tests missing

File: src/libraries/mysterium-client/monitoring/monitoring.js (hunk 3 above)
Comment (mipo47) [reply src/libraries/mysterium-client/monitoring/monitoring.js]: Suggestion:
```
return onFirstEventOrTimeout(this.onStatusUp, MYSTERIUM_CLIENT_WAITING_THRESHOLD)
```
to make it compile change Subcriber class: 
```
type Subscriber = (Callback) => any
```

File: src/app/events.js [hunk 7]
```diff
@@ -43,6 +43,7 @@ function onFirstEvent (subscriber: Subscriber): Promise<any> {
  *
  * @returns {Promise<any>}
  */
+// TODO: unsubscribe after timeout or event
 function onFirstEventOrTimeout (subscriber: Subscriber, timeout: number): Promise<void> {
```
Comment (mipo47) [inline src/app/events.js]: out of scope, but return type is incorrect, see resolve(data)

File: src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mipo47) [reply src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js]: if upgrade is timed out you will still got message:
logger.info('Upgrading: upgraded process is up')

Also this error is ignored:
throw new Error('Waiting for upgrade timed out')

File: src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js [hunk 8]
```diff
@@ -55,25 +64,60 @@ class LaunchDaemonProcess implements Process {
...
+    }
+    throw new Error('Waiting for upgrade timed out')
+  }
+
+  async _isUpgradeFinished (): Promise<boolean> {
+    try {
+      logger.info('Checking if upgrade finished')
+      let matches = await this._versionCheck.runningVersionMatchesPackageVersion()
+      if (matches) {
+        logger.info('Upgrade finished - versions matched')
```
[the remark this answers is not in the document]
Comment (donce) [reply src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js]: This seems a bit less readable to me - having mutable variable and assigning unmatched value by default.

File: src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (donce) [reply src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js]: >if upgrade is timed out you will still got message:
logger.info('Upgrading: upgraded process is up')

No, you will not see it, `finally` doesn't mean error is catched. If Error happens, `finally` block is executed first and then Error is propogated up, without invoking  `logger.info('Upgrading: upgraded process is up')`

File: src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js (hunk 2 above)
Comment (donce) [reply src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js]: Done

File: src/libraries/mysterium-client/monitoring/monitoring.js (hunk 3 above)
Comment (donce) [reply src/libraries/mysterium-client/monitoring/monitoring.js]: Fixed by using unsubscribers inside those util functions

File: src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js [hunk 9]
```diff
@@ -55,25 +64,60 @@ class LaunchDaemonProcess implements Process {
...
-    await this._monitoring.waitForStatusDownWithTimeout()
 
-    logger.info('Restarting: process is down, starting it up')
-    await this.start()
-    logger.info('Restarting: waiting for process to be up')
-    await this._monitoring.waitForStatusUpWithTimeout()
+    const unsubscribe = this._monitoring.onStatusDown(() => { this.start() })
+    try {
+      logger.info('Upgrading: waiting for upgraded client')
+      const UPGRADE_WAITING_TIMEOUT = 15000
```
[the remark this answers is not in the document]
Comment (donce) [reply src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js]: done

File: src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js [hunk 10]
```diff
@@ -55,25 +64,60 @@ class LaunchDaemonProcess implements Process {
...
+  }
+
+  async _waitForUpgrade (timeout: number) {
+    let timeFinished = false
+    setTimeout(() => {
+      timeFinished = true
+    }, timeout)
 
-    logger.info('Restarting: process is up')
+    const RETRY_DELAY = 200
```
[the remark this answers is not in the document]
Comment (donce) [reply src/libraries/mysterium-client/launch-daemon/launch-daemon-process.js]: done

File: test/unit/libraries/subscriber.spec.js [hunk 11]
```diff
@@ -70,4 +70,17 @@ describe('Subscriber', () => {
     expect(value1).to.eql('hey')
     expect(value2).to.eql('hey')
   })
+
+  it('returns function which unsubscribes', () => {
```
[the remark this answers is not in the document]
Comment (donce) [reply test/unit/libraries/subscriber.spec.js]: I throws exception - added spec for that.

File: src/libraries/mysterium-client/monitoring/monitoring.js (hunk 4 above)
Comment (donce) [reply src/libraries/mysterium-client/monitoring/monitoring.js]: Unrelated to this PR.

File: src/libraries/mysterium-client/monitoring/monitoring.js (hunk 5 above)
Comment (donce) [reply src/libraries/mysterium-client/monitoring/monitoring.js]: Unrelated to this PR.

File: src/libraries/mysterium-client/monitoring/monitoring.js (hunk 6 above)
Comment (donce) [reply src/libraries/mysterium-client/monitoring/monitoring.js]: Unrelated to this PR.

File: src/app/events.js (hunk 7 above)
Comment (donce) [reply src/app/events.js]: Fixed.

**Comment by donce** [comment]
@shroomist @mipo47 fixed.

File: src/app/mysterium-vpn.js [hunk 12]
```diff
@@ -317,9 +317,7 @@ class MysteriumVpn {
       htmlContent: this._terms.getContent()
     })
 
-    const termsAnsweredDTO = await onFirstEvent((callback) => {
-      this._communication.termsAnswered.on(callback)
-    })
+    const termsAnsweredDTO = await onFirstEvent(this._communication.termsAnswered.on.bind(this))
```
Comment (shroomist) [inline src/app/mysterium-vpn.js]: !!?? are you sure you're binding to the right object?
```suggestion
    const termsAnsweredDTO = await onFirstEvent(this._communication.termsAnswered.on.bind(this._communication.termsAnswered))
```

File: src/app/events.js [hunk 13]
```diff
@@ -29,9 +32,21 @@ type Subscriber = (Callback) => void
  */
 function onFirstEvent (subscriber: Subscriber): Promise<any> {
   return new Promise((resolve) => {
-    subscriber((data) => {
+    let eventReceived = false
+    let unsubscriber: ?Unsubscriber = null
+
+    unsubscriber = subscriber((data) => {
```
Comment (shroomist) [inline src/app/events.js]:
```suggestion
    unsubscriber = subscriber((data: any) => {
```

**Pull request merged**